sda-commons-server-auth/src/test/java/org/sdase/commons/server/auth/key/RsaKeyLoaderSchedulerTest.java

```diff
@@ -0,0 +1,70 @@
+package org.sdase.commons.server.auth.key;
+
+import static java.util.Collections.singletonList;
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.security.interfaces.RSAPublicKey;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import org.junit.After;
+import org.junit.Test;
+import org.mockito.Mockito;
+
+public class RsaKeyLoaderSchedulerTest {
+
+  private RsaPublicKeyLoader keyLoader = new RsaPublicKeyLoader();
+  private ScheduledExecutorService executorService = Executors.newSingleThreadScheduledExecutor();
+  private RsaKeyLoaderScheduler keyLoaderScheduler =
+      new RsaKeyLoaderScheduler(keyLoader, executorService);
+  private boolean includingKeyToBeRevoked = true;
+
+  @Test
+  public void shouldNotBeAuthenticatedWithRevokedKey() throws InterruptedException {
+    // given
+    final RSAPublicKey validKey = Mockito.mock(RSAPublicKey.class);
+    final String kidForValidKey = "kidForValidKey";
+    final RSAPublicKey keyToBeRevoked = Mockito.mock(RSAPublicKey.class);
+    final String kidForKeyToBeRevoked = "kidForKeyToBeRevoked";
+    final int period = 2;
+    final TimeUnit timeUnit = TimeUnit.SECONDS;
+
+    KeySource keySource =
+        new KeySource() {
+          @Override
+          public List<LoadedPublicKey> loadKeysFromSource() throws KeyLoadFailedException {
+            if (includingKeyToBeRevoked) {
+              return Arrays.asList(
+                  new LoadedPublicKey(kidForValidKey, validKey, this),
+                  new LoadedPublicKey(kidForKeyToBeRevoked, keyToBeRevoked, this));
+            } else {
+              return singletonList(new LoadedPublicKey("kidForValidKey", validKey, this));
+            }
+          }
+        };
+
+    keyLoader.addKeySource(keySource);
+    keyLoaderScheduler.start(0, period, timeUnit);
+    assertThat(keyLoader.getKey(kidForValidKey)).isSameAs(validKey);
+    assertThat(keyLoader.getKey(kidForKeyToBeRevoked)).isSameAs(keyToBeRevoked);
+
+    // when
+    // simulate revoking of key:
+    includingKeyToBeRevoked = false;
+
+    // wait for scheduler-execution:
+    timeUnit.sleep(period);
```

Comment:
Please use the awaitability package to use wait for the condition below to succeed. This should avoid flaky tests. Don't configure timeouts when using it, the defaults are fine. Like [here ](https://github.com/SDA-SE/sda-dropwizard-commons/blob/1b60c8ba20f00601e35ca91be2725b48d93ed135/sda-commons-server-opentracing/src/test/java/org/sdase/commons/server/opentracing/OpenTracingBundleTest.java#L65-L70).

It doesn't feel that it's guaranteed that the key load is completed after waiting for two seconds as the scheduler is also waiting two seconds, right?

Reply:
That's a cool tool. I didn't know it yet, thanks for the tip! Solves a lot of problems around testing asynchronous code.

Yes, two seconds maybe too short. I think the `await().untilAsserted` - statement is much more elegant for solving this.